### PR TITLE
receive: Include current config hash in forward requests

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

* Add the hashring hash to the forward payload.

## Verification

* `make test-local`

### CODE REVIEW

File: pkg/receive/handler.go [hunk 1]
```diff
@@ -50,6 +50,9 @@ const (
 	DefaultTenantLabel = "tenant_id"
 	// DefaultReplicaHeader is the default header used to designate the replica count of a write request.
 	DefaultReplicaHeader = "THANOS-REPLICA"
+	// DefaultHashringConfigHeader is the default header used to designate the hashring config hash of a write request.
+	DefaultHashringConfigHeader = "THANOS-HASHRING-CONFIG"
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply pkg/receive/handler.go:54]: I just scribbled these changes. So don't focus on the details I'll clean these parts. I might have unnecessarily add those parts.

File: pkg/receive/handler.go [hunk 2]
```diff
@@ -250,12 +255,16 @@ type replica struct {
...
 
-func (h *Handler) handleRequest(ctx context.Context, rep uint64, tenant string, wreq *prompb.WriteRequest) error {
+func (h *Handler) handleRequest(ctx context.Context, rep uint64, tenant string, config string, wreq *prompb.WriteRequest) error {
 	// The replica value in the header is one-indexed, thus we need >.
 	if rep > h.options.ReplicationFactor {
 		return errBadReplica
 	}
 
+	if h.hashring.ConfigHash() != config {
+		return errHashringConfigMismatch
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply pkg/receive/handler.go:265]: 👍 Alright. Make sense. It'd simplify things. I'm gonna add a metric and log this event.

File: pkg/receive/handler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/receive/handler.go:54]: Ack (sorry just got excited by the PR title :) )

File: pkg/receive/handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kakkoyun) [reply pkg/receive/handler.go]: If we won't fail until we run mutlple multi-tsdb's per hashring configuration (as it suggested in https://github.com/thanos-io/thanos/issues/3141), this is only going to be an informative not actionable.

**Comment by kakkoyun** [comment]
@brancz Do you see any value for merging this? Or should I just close this one?

**Comment by brancz** [comment]
I think since we're splitting the architecture for the receive component, we can close this. Wdyt?

**Comment by kakkoyun** [comment]
@brancz Sounds good to me 👍

**Pull request closed without merging**